sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/HashPartitionPushDownSupport.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.annotation.InterfaceStability;
+
+/**
+ * A mix in interface for `DataSourceV2Reader`. Users can implement this interface to pre-partition
+ * the data and avoid shuffle at Spark side.
+ *
+ * Note that this interface is marked as unstable, as the implementation needs to be consistent
```

Comment:
It seems like the intent here is to expose the underlying data's structure. For example, tables that were stored with Hive bucketing could return true if `partitionColumns` matches how the table was bucketed, and then reduce-side tasks could take advantage of that.

If that's the intent, then I would rather see an API that allows the data source report its structure, rather than a way for the planner to ask about a specific partitioning. Also, using the term "partition" here is somewhat confusing since we're talking about RDD partitions, not table partitions, right?